crates/compression/src/decompress.rs

```diff
@@ -0,0 +1,336 @@
+        Decompress,
+        DecompressibleBy,
+        RegistryKey,
+    },
+    fuel_tx::{
+        input::{
+            coin::{
+                Coin,
+                CoinSpecification,
+            },
+            message::{
+                Message,
+                MessageSpecification,
+            },
+            AsField,
+            PredicateCode,
+        },
+        AssetId,
+        CompressedUtxoId,
+        Mint,
+        ScriptCode,
+        Transaction,
+        UtxoId,
+    },
+    fuel_types::{
+        Address,
+        ContractId,
+    },
+    tai64::Tai64,
+};
+
+pub trait DecompressDb: TemporalRegistryAll + HistoryLookup {}
+impl<T> DecompressDb for T where T: TemporalRegistryAll + HistoryLookup {}
+
+/// This must be called for all decompressed blocks in sequence, otherwise the result will be garbage.
+pub async fn decompress<D>(
+    config: Config,
+    mut db: D,
+    block: VersionedCompressedBlock,
+) -> anyhow::Result<PartialFuelBlock>
+where
+    D: DecompressDb,
+{
+    let VersionedCompressedBlock::V0(compressed) = block;
+
+    // TODO: merkle root verification: https://github.com/FuelLabs/fuel-core/issues/2232
+
+    compressed
+        .registrations
+        .write_to_registry(&mut db, compressed.header.consensus.time)?;
+
+    let ctx = DecompressCtx {
+        config,
+        timestamp: compressed.header.consensus.time,
+        db,
+    };
+
+    let transactions = <Vec<Transaction> as DecompressibleBy<_>>::decompress_with(
+        compressed.transactions,
+        &ctx,
+    )
+    .await?;
+
+    Ok(PartialFuelBlock {
+        header: compressed.header,
+        transactions,
+    })
+}
+
+pub struct DecompressCtx<D> {
+    pub config: Config,
+    /// Timestamp of the block being decompressed
+    pub timestamp: Tai64,
+    pub db: D,
+}
+
+impl<D> ContextError for DecompressCtx<D> {
+    type Error = anyhow::Error;
+}
+
+impl<D> DecompressibleBy<DecompressCtx<D>> for UtxoId
+where
+    D: HistoryLookup,
+{
+    async fn decompress_with(
+        c: CompressedUtxoId,
+        ctx: &DecompressCtx<D>,
+    ) -> anyhow::Result<Self> {
+        ctx.db.utxo_id(c)
+    }
+}
+
+macro_rules! decompress_impl {
+    ($($type:ty),*) => { paste::paste! {
+        $(
+            impl<D> DecompressibleBy<DecompressCtx<D>> for $type
+            where
+                D: TemporalRegistry<$type>
+            {
+                async fn decompress_with(
+                    key: RegistryKey,
+                    ctx: &DecompressCtx<D>,
+                ) -> anyhow::Result<Self> {
+                    if key == RegistryKey::DEFAULT_VALUE {
+                        return Ok(<$type>::default());
+                    }
+                    let key_timestamp = ctx.db.read_timestamp(&key)?;
+                    if !ctx.config.is_timestamp_accessible(ctx.timestamp, key_timestamp)? {
+                        anyhow::bail!("Timestamp not accessible");
+                    }
+                    ctx.db.read_registry(&key)
```

Comment:
We can do that in a separate PR, but it would be nice to have internal cache in the context to avoid reading data from the database. Even it in the RocksDB cache, we still spend time on deserialization.

Reply:
My first intuition says that the cache should live on the thing that implements the trait, and not here. I also don't like the idea of caching without first measuring that this is indeed a performance issue.